hoomd/md/tune/nlist_buffer.py

```diff
@@ -0,0 +1,213 @@
+# Copyright (c) 2009-2022 The Regents of the University of Michigan.
+# Part of HOOMD-blue, released under the BSD 3-Clause License.
+
+"""Provide a tuner for `hoomd.md.nlist.NeighborList.buffer`."""
+
+import copy
+
+import numpy as np
+
+import hoomd.custom
+import hoomd.data
+from hoomd.data.typeconverter import OnlyTypes, SetOnce
+import hoomd.logging
+import hoomd.tune
+import hoomd.trigger
+from hoomd.md.nlist import NeighborList
+
+
+class _IntervalTPS:
+
+    def __init__(self, simulation):
+        self._simulation = simulation
+        self._last_timestep = simulation.timestep
+        self._last_walltime = simulation.walltime
+        self._last_tps = None
+
+    def __call__(self):
+        if self._simulation.timestep == self._last_timestep:
+            return self._last_tps
+        walltime = self._simulation.walltime
+        timestep = self._simulation.timestep
+        if self._last_walltime is not None and self._last_timestep is not None:
+            delta_w = walltime - self._last_walltime
+            delta_t = timestep - self._last_timestep
+            # We divide by 1_000 to reduce the gradient size for optimization.
+            tps = delta_t / (1e3 * delta_w)
+        else:
+            tps = None
+        self._last_tps = tps
+        self._last_walltime = walltime
+        self._last_timestep = timestep
+        return tps
+
+
+class _NeighborListBufferInternal(hoomd.custom._InternalAction):
+    _skip_for_equality = {"_simulation", "_tunable"}
+
+    def __init__(
+        self,
+        nlist: NeighborList,
+        solver: hoomd.tune.solve.Optimizer,
+        maximum_buffer: float,
+    ):
+        self._simulation = None
+        self._tuned = False
+        self._max_tps = 0
+        self._best_buffer_size = None
+        param_dict = hoomd.data.parameterdicts.ParameterDict(
+            nlist=SetOnce(NeighborList),
+            solver=SetOnce(hoomd.tune.solve.Optimizer),
+            maximum_buffer=OnlyTypes(float,
+                                     postprocess=self._maximum_buffer_post))
+        param_dict.update({
+            "nlist": nlist,
+            "solver": solver,
+            "maximum_buffer": maximum_buffer
+        })
+        self._param_dict.update(param_dict)
+
+    def act(self, timestep: int):
+        if not self.tuned:
+            tps = self._tunable.y
+            if tps is not None and tps > self._max_tps:
+                self._best_buffer_size = self._tunable.x
+                self._max_tps = tps
+            self._tuned = self.solver.solve([self._tunable])
+
+    def _maximum_buffer_post(self, value: float):
+        if self._simulation is not None:
+            self._tunable.domain = (1e-5, value)
```

Comment:
```suggestion
            self._tunable.domain = (0, value)
```
Users may run in a system of units where 1e-5 is very large value for buffer.